### PR TITLE
Fix HydrogenBondAnalysis when no atom pair is with 3A

### PR DESCRIPTION
**Pull request by xiki-tempula**
When there is no atom pair between the two selections of HydrogenBondAnalysis that are within 3A, HydrogenBondAnalysis throws an exception instead of giving a empty list.
Code to reproduce the error
```
import MDAnalysis as mda
import MDAnalysis.analysis.hbonds

text = '''TITLE     Two atoms far away
CRYST1   52.763   52.763   52.763  90.00  90.00  90.00 P 1           1
MODEL         1
ATOM      1  N   LEU     1      32.310  13.778  14.372  1.00  0.00      SYST N 0
ATOM      2  OW  SOL     2       3.024   4.456   4.147  1.00  0.00      SYST H 0'''
with open('temp.pdb', 'w') as f:
    f.write(text)
u = mda.Universe('temp.pdb')
h = MDAnalysis.analysis.hbonds.HydrogenBondAnalysis(u, 'resname SOL', 'protein')
h.run()
```

Gives `AttributeError: 'list' object has no attribute 'select_atoms'`

After analysing the code, it seems that there is no check after filtering to make sure that a selection of atom can be done. 

This PR fixed this problem.

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by orbeckst** [comment]
@xiki-tempula thanks for report and fix. 

You probably need to rebase your branch against develop and force push.

I also don't understand why travis did not run; there should be a proper unit test run. Hopefully that should show up after you push again.

As a minor thing: If you find a bug, raise an issue and describe the bug (as you did in the first part of your report). *Then* open a PR that *references* the issue. The reason is that sometimes PRs get thrown away or changed. Then you are also loosing a good reference point. It's generally considered good practice to separate the issue report from the PR (or PRs). No big deal, but for the future.

File: package/CHANGELOG [hunk 1]
```diff
@@ -13,15 +13,26 @@ The rules for this file:
...
-mm/dd/17 utkbansal, kain88-de
+
+mm/dd/17 utkbansal, kain88-de, xiki-tempula, kaplajon
+
 
   * 0.16.1
 
 Enhancements
   * Universe now works with StringIO objects for topologies and trajectories.
+  * Residues with the same residue ids are not merged by default now
```
Comment (orbeckst) [inline package/CHANGELOG]: There are lots of pieces that should not show up in your PR, they come from elsewhere. Something went wrong when you merged. Rebase against develop.

File: package/MDAnalysis/analysis/hbonds/hbond_analysis.py [hunk 2]
```diff
@@ -816,20 +816,21 @@ def _update_selection_2(self):
...
-            self._s2_donors = self._s2.select_atoms(
-                'name {0}'.format(' '.join(self.donors)))
-            self._s2_donors_h = {}
-            for i, d in enumerate(self._s2_donors):
-                tmp = self._get_bonded_hydrogens(d)
-                if tmp:
-                    self._s2_donors_h[i] = tmp
-            self.logger_debug("Selection 2 donors: {0:d}".format(len(self._s2_donors)))
-            self.logger_debug("Selection 2 donor hydrogens: {0:d}".format(len(self._s2_donors_h)))
+        if self._s2:
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: Don't you need a similar check for `_s1`? 

You can probably re-use your test case by just switching selections.

Comment (orbeckst) [inline package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: In my opinion, it would be cleaner to just return here instead of doing a big `if` block. If you re-order the code so that all data structures are initialized and then use the `if not self._s2` where the logger is already reporting  this problem and `return` from there.

```python
         self._s2_donors = {}
         self._s2_donors_h = {}
         self._s2_acceptors = {}

         if not self._s2:
            # nothing to do here, just keep empty data structures
            logger.warn('Selection 2 "{0}" did not select any atoms.'.format(
                 str(self.selection2)[:80]))
            return

        if self.selection1_type in ('donor', 'both'):
            self._s2_acceptors = self._s2.select_atoms(...)
            ...
```

This is reasonably clean (and the diff will also be smaller).

**Comment by xiki-tempula** [comment]
@orbeckst Thank you for your advice.
I have checked and it seems that this bug occurred because there was no check after filtering.
There are checks at the beginning which dealt with an empty selection. Then the script compares the pairwise distance between s1 and s2 and then get rid of the atoms in s2 which is too far away from s1. Thus, s2 is changed while s1 is not. So if the s2 becomes empty after filtering, it escapes the check and caused error.
This can be check using
h = MDAnalysis.analysis.hbonds.HydrogenBondAnalysis(u, 'resname SOL', 'protein and not protein')

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 3]
```diff
@@ -93,6 +88,18 @@ def test_generate_table(self):
         assert_array_equal(h.table.donor_resid, self.values['donor_resid'])
         assert_array_equal(h.table.acceptor_resnm, self.values['acceptor_resnm'])
 
+    def test_atoms_too_far(self):
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Minor thing: Make this a static method
```python
 @staticmethod
def test_atoms_too_far():
   ...
```

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 4]
```diff
@@ -31,8 +31,12 @@
 
 import itertools
 import warnings
+try:
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Why is this needed?? I think it can be removed, can't it?

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 5]
```diff
@@ -87,6 +91,18 @@ def test_generate_table(self):
         assert_array_equal(h.table.donor_resid, self.values['donor_resid'])
         assert_array_equal(h.table.acceptor_resnm, self.values['acceptor_resnm'])
 
+    def test_atoms_too_far(self):
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Minor thing: Make this a static method
```python
@staticmethod
def test_atoms_too_far():
   ...
```

File: testsuite/MDAnalysisTests/data/HBond_atoms_far.pdb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+TITLE     Two atoms far away
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/data/HBond_atoms_far.pdb]: remove (you're using the inline test)

File: testsuite/MDAnalysisTests/datafiles.py [hunk 7]
```diff
@@ -144,6 +144,7 @@
     "MMTF", "MMTF_gz",
     "ALIGN_BOUND",  # two component bound system
     "ALIGN_UNBOUND", # two component unbound system
+    "HBond_atoms_far", # Universe with two atoms far away
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/datafiles.py]: remove (you're using the inline test)

File: testsuite/MDAnalysisTests/datafiles.py [hunk 8]
```diff
@@ -221,6 +222,7 @@
 XTC_sub_sol = resource_filename(__name__, 'data/cobrotoxin.xtc')
 PDB_sub_sol = resource_filename(__name__, 'data/cobrotoxin.pdb')
 PDB_xlserial = resource_filename(__name__, 'data/xl_serial.pdb')
+HBond_atoms_far = resource_filename(__name__, 'data/HBond_atoms_far.pdb')
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/datafiles.py]: remove (you're using the inline test)

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 9]
```diff
@@ -88,6 +92,18 @@ def test_generate_table(self):
         assert_array_equal(h.table.acceptor_resnm, self.values['acceptor_resnm'])
 
     @staticmethod
+    def test_atoms_too_far(self):
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: remove `self` from the argument list of the static method

**Comment by orbeckst** [comment]
I resolved the conflict in CHANGELOG using GitHub's friendly *Resolve* button (I hadn't done this before). Apparently that merged develop into your branch. Normally that is not what I would do but this should be ok. Remember to `git pull` before committing to this branch.

When it comes to merging I will probably do any squashing manually from the command line, just to make sure that the merge of develop into your branch did not mess up anything.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 4 above)
Comment (xiki-tempula) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Thank you for your comment. There is no StringIO in py3 so I was trying to make it compatible with py3. I still need some time to figure out how this stream reading works.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 10]
```diff
@@ -103,7 +104,6 @@ def test_atoms_too_far(self):
         h.run(verbose=False)
         assert_equal(h.timeseries, [[]])
 
-    @staticmethod
     def test_true_traj():
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: This `@staticmethod` has to stay.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 4 above)
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Oh, I didn't think about this. Then we should probably use [six.StringIO](https://pythonhosted.org/six/#six.StringIO): something like
```python
from six import StringIO
```

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 11]
```diff
@@ -88,6 +92,18 @@ def test_generate_table(self):
...
+MODEL         1
+ATOM      1  N   LEU     1      32.310  13.778  14.372  1.00  0.00      SYST N 0
+ATOM      2  OW  SOL     2       3.024   4.456   4.147  1.00  0.00      SYST H 0'''
+
+        u = MDAnalysis.Universe(BytesIO(pdb.encode()), format="pdb")
+        h = MDAnalysis.analysis.hbonds.HydrogenBondAnalysis(u, 'resname SOL', 'protein')
+        h.run(verbose=False)
+        assert_equal(h.timeseries, [[]])
+
     def test_true_traj():
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: There has to be a `@staticmethod` in front of the `test_true_traj()` function.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 12]
```diff
@@ -31,8 +31,12 @@
 
 import itertools
 import warnings
+try:
+    from BytesIO import BytesIO
+except ImportError:
+    from io import BytesIO
 
-from MDAnalysisTests.datafiles import PDB_helix, GRO, XTC
+from MDAnalysisTests.datafiles import PDB_helix, GRO, XTC, HBond_atoms_far
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: remove HBond_atoms_far

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 13]
```diff
@@ -88,6 +92,18 @@ def test_generate_table(self):
...
+CRYST1   52.763   52.763   52.763  90.00  90.00  90.00 P 1           1
+MODEL         1
+ATOM      1  N   LEU     1      32.310  13.778  14.372  1.00  0.00      SYST N 0
+ATOM      2  OW  SOL     2       3.024   4.456   4.147  1.00  0.00      SYST H 0'''
+
+        u = MDAnalysis.Universe(BytesIO(pdb.encode()), format="pdb")
+        h = MDAnalysis.analysis.hbonds.HydrogenBondAnalysis(u, 'resname SOL', 'protein')
+        h.run(verbose=False)
+        assert_equal(h.timeseries, [[]])
+
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: There has to be a @staticmethod in front of the test_true_traj() function; Travis fails https://travis-ci.org/MDAnalysis/mdanalysis/jobs/227170225 , too, and QC flags it as a [critical issue](https://www.quantifiedcode.com/app/project/gh:MDAnalysis:mdanalysis/diff/401c36636a5c12d250d2edd08ef44f678fe8ba45/2804d48f5bcc914ed461f64ba80e96c8ffb9a135/issues?groups=code_patterns%3A3bECxdfc).

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 4 above)
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Use `six.StringIO` (as discussed in previous comment)

**Comment by xiki-tempula** [comment]
@orbeckst Thank you for your advice. I guess I have learnt about decorators today.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 4 above)
Comment (xiki-tempula) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Thank you. It seems that six is more powerful than I thought after spending an hour playing with py3.io.StringIO.

**Comment by orbeckst** [comment]
I am rebasing/squashing/merging this manually.

**Comment by orbeckst** [comment]
See PR #1327 (I needed to make a new PR for the rebasing/history rewriting).

**Comment by xiki-tempula** [comment]
@orbeckst I have approved it. Thank you for all the help and advice.

**Comment by orbeckst** [comment]
Thanks for fixing a bug! 

We'll have to wait for another coredev to approve #1327 – I can't do it because I submitted the PR.
